0015-estimator-interface.md

```diff
@@ -0,0 +1,457 @@
+
+```python
+TaskResult({“evs”: <20, 30>, “stds”: <20, 30>}, metadata)
+```
+
+The reason we are proposing a generic container for the return type instead of, e.g., an `Estimator`-specific container, is because it
+
+1. Provides unified experience across primitives for users. 
+2. code-reuse
+3. Provides a certain certain amount of flexibility for what can be returned without modifying the container object. Here are some examples:
+  1. Suppose that we want to give users the option of additionally returning the covariances between estimates that arise because of the circuit multiplexing, then we could update with the field `{"cov": <20,30,20,30>}`.
+  2. Suppose we want to return some indication of which estimates came from the same physical circuit.
+
+## Migration Path <a name="migration-path"></a>
+
+We need to remain backwards compatible with the existing interface to adhere to the [Qiskit Deprecation Policy](https://qiskit.org/documentation/deprecation_policy.html).
+
+In summary, we propose the following strategy: If the user has provided no `TaskLike`s, proceed with the old API, the old API output, and emit a deprecation warning, or an error if something mandatory like `observables` has been omitted. Otherwise, proceed with the new API, raising if they have tried to use the old arguments in addition to providing tasks.
+
+We now describe the strategy in detail, beginning with a restatement of the current interface, where `Estimator.run` has `circuits` as the only positional argument and accepts `observables` and `parameter_values` as keyword arguments (in addition to the var keyword `**run_options` arguments which isn't affected):
+
+```python
+class Estimator(BasePrimitive):
+
+    # current signature
+    def run(
+            self,
+            circuits: Sequence[QuantumCircuit] | QuantumCircuit,
+            observables: Sequence[BaseOperator | PauliSumOp | str] | BaseOperator | PauliSumOp | str,
+            parameter_values: Sequence[Sequence[float]] | Sequence[float] | float | None = None,
+            **run_options,
+        ):
+        ...
+```
+
+We propose the migration to using tasks in two phases.
+
+### Deprecation Phase 1
+
+In the first phase:
+* Introduce `tasks: Sequence[EstimatorTask] | EstimatorTask` as the only positional argument, and make `circuits` a keyword argument.
+* Coerce the arguments to account for the fact that the only existing positional argument is `circuit: Sequence[QuantumCircuit] | QuantumCircuit`.
+* Check that the user does not attempt to mix the old/new APIs.
+* If necessary, coerce the old-API arguments (now all keyword arguments) into `EstimatorTask`s, **raise deprecation warning**.
+* Always eventually run with `Estimator._run(tasks: Sequence[EstimatorTask], **run_options)`.
+
+Here is a mock implementation, with missing type annotations as above:
+```python
+class Estimator(BasePrimitive):
+
+    def run(
+        self, 
+        tasks: Sequence[EstimatorTask] | EstimatorTask, 
```

Comment:
A minor detail, but we'd have move `tasks` to the end for backward compatibility.